core/src/main/scala/org/apache/spark/h2o/converters/H2ORDD.scala

```diff
@@ -58,63 +62,166 @@ class H2ORDD[A <: Product: TypeTag: ClassTag, T <: Frame] private(@transient val
   }
 
   /** Number of columns in the full dataset */
-  val numCols = frame.numCols()
-  val types = ReflectionUtils.types[A](colNames)
+  val numColsInFrame = frame.numCols()
+
+  val colNamesInFrame = frame.names()
+
+  val types = ReflectionUtils.types[A](colNamesInResult)
   val expectedTypesAll: Option[Array[Byte]] = ConverterUtils.prepareExpectedTypes(isExternalBackend, types)
 
   /**
    * :: DeveloperApi ::
    * Implemented by subclasses to compute a given partition.
    */
   override def compute(split: Partition, context: TaskContext): Iterator[A] = {
+    new H2ORDDIterator(frameKeyName, split.index)
+  }
 
-    val iterator = new H2OChunkIterator[A] {
-
-      val jc = implicitly[ClassTag[A]].runtimeClass
-      val cs = jc.getConstructors
-      val ccr = cs.collectFirst({
-                case c if c.getParameterTypes.length == colNames.length => c
-              })
-        .getOrElse({
-            throw new IllegalArgumentException(
-                  s"Constructor must take exactly ${colNames.length} args")
-      })
-
-      override val keyName = frameKeyName
-      override val partIndex = split.index
-      val expectedTypes: Option[Array[Byte]] = expectedTypesAll
-
-      /* Converter context */
-      override val converterCtx: ReadConverterContext =
-      ConverterUtils.getReadConverterContext(isExternalBackend,
-        keyName,
-        chksLocation,
-        expectedTypes,
-        partIndex)
+  private val columnTypeNames = ReflectionUtils.typeNames[A](colNamesInResult)
+
+  private val jc = implicitly[ClassManifest[A]].runtimeClass
+
+  private type Deserializer = ReadConverterContext => Int => Any
```

Comment:
New types should be introduced. They specify exactly what we mean. Throwing in tuples and strings and arrays is not a good Scala practice.

Regarding the word `Reader`, it's a nice word, but it's way too popular in a general context. Are we dealing with a Reader monad? Most probably not.

As to parameterizing the serializer, it would be nice, and there's a lot of reason to do that, but I do not see how it is possible without using `HList`. The extractor map will be heterogeneous. Anyway, will think about it.